kernel/posix/pthread.c

```diff
@@ -0,0 +1,537 @@
+
+	return prio;
+}
+
+static s32_t posix_to_zephyr_priority(u32_t priority, int policy)
+{
+	s32_t prio;
+
+	if (policy == SCHED_FIFO) {
+		/* Zephyr COOP priority starts from -1 */
+		prio =  -1 * (priority + 1);
+	} else {
+		prio = (CONFIG_NUM_PREEMPT_PRIORITIES - priority);
+	}
+
+	return prio;
+}
+
+/**
+ * @brief Set scheduling parameter attributes in thread attributes object.
+ *
+ * See IEEE 1003.1
+ */
+int pthread_attr_setschedparam(pthread_attr_t *attr,
+			       const struct sched_param *schedparam)
+{
+	int priority = schedparam->priority;
+
+	if (!attr || !attr->initialized) {
+		return EINVAL;
+	}
+
+	if (is_posix_prio_valid(priority, attr->schedpolicy) == false) {
+		return ENOTSUP;
+	}
+
+	attr->priority = priority;
+	return 0;
+}
+
+/**
+ * @brief Set stack attributes in thread attributes object.
+ *
+ * See IEEE 1003.1
+ */
+int pthread_attr_setstack(pthread_attr_t *attr, void *stackaddr,
+			  size_t stacksize)
+{
+	if (stackaddr == NULL) {
+		return EACCES;
+	}
+
+	attr->stack = stackaddr;
+	attr->stacksize = stacksize;
+	return 0;
+}
+
+static void zephyr_thread_wrapper(void *arg1, void *arg2, void *arg3)
+{
+	void * (*fun_ptr)(void *) = arg3;
+
+	fun_ptr(arg1);
+	pthread_exit(NULL);
+}
+
+/**
+ * @brief Create a new thread.
+ *
+ * See IEEE 1003.1
+ */
+int pthread_create(pthread_t *newthread, const pthread_attr_t *attr,
+		   void *(*threadroutine)(void *), void *arg)
+{
+	static atomic_val_t thread_count;
+	atomic_val_t thread_index = 0;
+	s32_t prio;
+	pthread_mutexattr_t att;
+	pthread_condattr_t cond_attr;
+	struct posix_thread *thread;
+
+	if (!attr || !attr->initialized || !attr->stack || !attr->stacksize) {
+		return EINVAL;
+	}
+
+	prio = posix_to_zephyr_priority(attr->priority, attr->schedpolicy);
+	thread_index = atomic_inc(&thread_count);
+
+	/* FIXME: Thread space allocation need to use mempool for better
+	 * space utilization under limited number of threads.
+	 */
+	if (thread_index < CONFIG_MAX_PTHREAD_COUNT) {
+		thread = &pthread[thread_index];
+		thread->cancel_state = (1 << PTHREAD_CANCEL_POS) & attr->flags;
+		thread->state = attr->detachstate;
+		thread->cancel_pending = 0;
+		thread->state_lock.sem = &thread->state_lock_sem;
+		pthread_mutex_init(&thread->state_lock, &att);
+		thread->cancel_lock.sem = &thread->cancel_lock_sem;
+		pthread_mutex_init(&thread->cancel_lock, &att);
+		pthread_cond_init(&thread->state_cond, &cond_attr);
+
+		*newthread = (pthread_t)
+				k_thread_create(&thread->thread,
+						attr->stack, attr->stacksize,
+						(k_thread_entry_t)
+						zephyr_thread_wrapper,
+						(void *)arg, NULL,
+						threadroutine, prio,
+						attr->flags,
+						attr->delayedstart);
+	} else {
+		/* Insuffecient resource, maximum thread count reached */
+		return EAGAIN;
+	}
+
+	return 0;
+}
+
+
+/**
+ * @brief Set cancelability State.
+ *
+ * See IEEE 1003.1
+ */
+int pthread_setcancelstate(int state, int *oldstate)
+{
+	struct posix_thread *pthread = (struct posix_thread *) pthread_self();
```

Comment:
pthread_self() is currently just returning whatever k_current_get() returns. That's fine, but if the current thread isn't a pthread, the type punning is going to be undefined behavior and this might lead to memory corruption if it's called from a thread that's not been created with pthread_create(). I suggest implementing something like "_is_thread_pthread()", similar to "_is_in_isr()", and return an error before proceeding.

Reply:
hmm

But we are calling pthread_self() only from a POSIX. So it will always return a thread id which is corresponding to a POSIX thread.
 
Do you want to this implemented through creating a entry (something like pthread_valid into pthreadObject) or through some other way?

Reply:
This function can still be called from a thread not created by pthread_create().